src/mono/mono/mini/aot-runtime.c

```diff
@@ -4920,14 +4920,35 @@ mono_aot_get_method (MonoMethod *method, MonoError *error)
 	MonoClass *klass = method->klass;
 	MonoMethod *orig_method = method;
 	guint32 method_index;
-	MonoAotModule *amodule = m_class_get_image (klass)->aot_module;
+	MonoImage *image = m_class_get_image (klass);
 	guint8 *code = NULL;
 	gboolean cache_result = FALSE;
 	ERROR_DECL (inner_error);
 
 	error_init (error);
 
-	if (!amodule)
+	if (!(image->aot_module)) {
+		// aot_module was uninitialized
+		MonoMethodHeader *header = mono_method_get_header_checked (method, error);
+		if (!header) {
+			return NULL;
+		} else if (header->code_size != 0) {
+			return NULL;
+		} else {
+			// IL code for the method body doesn't exist. Try waiting for aot_module to be loaded probably by another thread
+			int count = 0;
+			while (!(image->aot_module) && (count < 10)) {
+				g_usleep (100);
+				count++;
```

Comment:
This isn't 100% correct (if the current thread is holding the loader lock, while another thread is trying to load some assemblies while loading hte AOT module, the other thread won't make progress).  But it should handle 80% of the cases (where we just had a simple race, rather than a potential deadlock).